File: 17-Binary-Tree/python/09_pre_post_inorder_in_one_traversal.py
class Solution:
    def treeTraversal(self, root):
        inorder = []
        preorder = []
        postorder = []

        def traverse(node):
            if not node:
                return
            preorder.append(node.data)
            traverse(node.left)
            inorder.append(node.data)
            traverse(node.right)
            postorder.append(node.data)

        traverse(root)
        return [inorder, preorder, postorder]

File: 17-Binary-Tree/cpp/09-pre-post-inorder-in-one-traversal.cpp
class Solution {
public:
    vector<vector<int>> treeTraversal(TreeNode* root) {
        vector<int> inorder, preorder, postorder;
        traverse(root, inorder, preorder, postorder);

        return {inorder, preorder, postorder};
    }

private:
    void traverse(TreeNode* node, vector<int>& inorder, vector<int>& preorder, vector<int>& postorder) {
        if (node == nullptr) return;

        preorder.push_back(node->data);
        traverse(node->left, inorder, preorder, postorder);
        inorder.push_back(node->data);
        traverse(node->right, inorder, preorder, postorder);
        postorder.push_back(node->data);
    }
};
